File: C/Linked List/Singly Linked List/SinglyLL010.c
/*
=====================================================================

Program Name    : Singly Linked List - Traversal Using Temporary Pointer
Program Number  : 010
Description     : This program demonstrates the use of a temporary
                  pointer for linked list traversal. Instead of
                  directly using the head pointer, a temporary
                  pointer is assigned the address of the first node
                  and is used to access the linked list nodes.

Concepts Used   : Structure, Self-Referential Structure,
                  typedef, Head Pointer,
                  Temporary Pointer, Node Traversal

Input           : None

Output          : Displays the data stored in all linked nodes.

Date            : July 2026
Language        : C
Author          : Atharv Tushar Bhosale

=====================================================================
*/

#include<stdio.h>

#pragma pack(1)

struct node
{
    int data;
    struct node *next;
};

typedef struct node NODE;
typedef struct node *PNODE;

int main()
{
    PNODE head = NULL;
    PNODE temp = NULL;
    NODE obj1, obj2, obj3;

    head = &obj1;

    obj1.data = 11;
    obj1.next = &obj2;

    obj2.data = 21;
    obj2.next = &obj3;

    obj3.data = 51;
    obj3.next = NULL;

    temp = head;

    printf("%d\n", temp->data);
    printf("%d\n", temp->next->data);
    printf("%d\n", temp->next->next->data);

    return 0;
}

/*
=====================================================================

Expected Output :
11
21
51

Explanation :
A temporary pointer named temp is created and
assigned the address stored in head. The temp
pointer is then used to access the linked nodes.

This approach preserves the head pointer while
allowing traversal of the linked list.

Operations Performed :
1. Created three linked nodes.
2. Created a head pointer.
3. Created a temporary pointer.
4. Assigned head to temp.
5. Accessed node data through temp.
6. Preserved the original head pointer.

Conclusion :
A temporary pointer is commonly used during linked
list traversal because it allows navigation through
the list without modifying the head pointer.

=====================================================================
*/

File: Java/Linked List/Singly Linked List/SinglyLL010.java
/*
=====================================================================

Name            : Singly Linked List - InsertFirst with Node Counter
Program Number  : 010

Description     : This program demonstrates the InsertFirst()
                  operation in a Singly Linked List along with
                  maintaining a node count. Each insertion at
                  the beginning updates the head pointer and
                  increments the node counter.

Concepts Used   : Class, Object Creation, Constructor,
                  Self-Referential Class,
                  InsertFirst Operation,
                  Head Pointer Manipulation,
                  Node Counter Management

Input           : 51, 21, 11

Output          : Nodes inserted successfully at the
                  beginning of the linked list

Date            : July 2026
Language        : Java
Author          : Atharv Tushar Bhosale

=====================================================================
*/

class node
{
    public int data;
    public node next;

    node(int no)
    {
        this.data = no;
        this.next = null;
    }
}

class SinglyLL
{
    private node first;
    private int iCount;

    public SinglyLL()
    {
        this.first = null;
        this.iCount = 0;
    }

    public void Display()
    {}

    public int Count()
    {
        return iCount;
    }

    public void InsertFirst(int iNo)
    {
        node newn = new node(iNo);

        if(first == null)
        {
            first = newn;
        }
        else
        {
            newn.next = first;
            first = newn;
        }

        iCount++;
    }

    public void InsertLast(int iNo)
    {
        node newn = new node(iNo);
        node temp = null;

        newn.data = iNo;
        newn.next = null;

        if(first == null)
        {
            first = newn;
        }
        else
        {

        }

        iCount++;
    }

    public void InsertAtPos(int iNo,int iPos)
    {}

    public void DeleteFirst()
    {}

    public void DeleteLast()
    {}

    public void DeleteAtPos(int iPos)
    {}
}

class SinglyLL010
{
    public static void main(String A[])
    {
        SinglyLL sobj = new SinglyLL();

        sobj.InsertFirst(51);
        sobj.InsertFirst(21);
        sobj.InsertFirst(11);
    }
}

/*
=====================================================================

Summary         : Implemented InsertFirst() operation with
                  automatic node count maintenance.

Linked List     : 11 -> 21 -> 51 -> NULL

Node Count      : 3

Learning        : Insertion at the beginning is the most
                  efficient linked list operation as it
                  requires only head pointer modification.

=====================================================================
*/
